distributed/shuffle/shuffle_extension.py

```diff
@@ -134,124 +213,78 @@ def inputs_done(self) -> None:
     def done(self) -> bool:
         return self.transferred and self.output_partitions_left == 0
 
+    def close(self):
+        self.multi_file.close()
+
 
 class ShuffleWorkerExtension:
     "Extend the Worker with routes and state for peer-to-peer shuffles"
 
     def __init__(self, worker: Worker) -> None:
         # Attach to worker
         worker.handlers["shuffle_receive"] = self.shuffle_receive
-        worker.handlers["shuffle_init"] = self.shuffle_init
         worker.handlers["shuffle_inputs_done"] = self.shuffle_inputs_done
         worker.extensions["shuffle"] = self
 
         # Initialize
         self.worker: Worker = worker
         self.shuffles: dict[ShuffleId, Shuffle] = {}
+        self.executor = ThreadPoolExecutor(worker.nthreads)
 
     # Handlers
     ##########
     # NOTE: handlers are not threadsafe, but they're called from async comms, so that's okay
 
-    def shuffle_init(self, comm: object, metadata: ShuffleMetadata) -> None:
-        """
-        Hander: Register a new shuffle that is about to begin.
-        Using a shuffle with an already-known ID is an error.
-        """
-        if metadata.id in self.shuffles:
-            raise ValueError(
-                f"Shuffle {metadata.id!r} is already registered on worker {self.worker.address}"
-            )
-        self.shuffles[metadata.id] = Shuffle(metadata, self.worker)
+    def heartbeat(self):
+        return {id: shuffle.heartbeat() for id, shuffle in self.shuffles.items()}
 
-    def shuffle_receive(
+    async def shuffle_receive(
         self,
         comm: object,
         shuffle_id: ShuffleId,
-        output_partition: int,
-        data: pd.DataFrame,
+        data: list[bytes],
     ) -> None:
         """
         Hander: Receive an incoming shard of data from a peer worker.
         Using an unknown ``shuffle_id`` is an error.
         """
-        self._get_shuffle(shuffle_id).receive(output_partition, data)
-
-    def shuffle_inputs_done(self, comm: object, shuffle_id: ShuffleId) -> None:
+        shuffle = await self._get_shuffle(shuffle_id)
+        future = asyncio.ensure_future(shuffle.receive(data))
```

Comment:
Yeah, that's a valid concern.  I don't yet have a great answer to this yet.  We can raise the exception here early if we want to (avoids hiding the exceptions that we're intentionally raising, but doesn't handle the unknown ones).  

If you're cool with it, I'd like to put in a TODO here so that we don't forget, and then defer this to future work.